File: p1/src/address.py
# Michael LeMay, 1/19/19

import numpy as np
import math

class Address:


    def __init__(self, newAddr, addrBits, numSets, numBlocks):
        """Initiates an address and all of its parameters. Parameters
        are needed in order to know the size of each field of the address.
        """
        self.addr = newAddr
        self.addrBits = addrBits;
        self.numSets = numSets;
        self.numBlocks = numBlocks;

        self.numBlockOffsetBits = self.calcNumBlockOffsetBits()
        self.numSetBits = self.calcNumSetBits()
        self.numTagBits = self.calcNumTagBits()


    def getTag(self):
        """Returns the tag bits with bitmasking"""
        mask = (1 << self.numTagBits) - 1
        mask = mask << (self.numSetBits + self.numBlockOffsetBits)
        return (self.addr & mask) >> (self.numSetBits + self.numBlockOffsetBits)


    def getIndex(self):
        """Returns the index bits with bitmasking"""
        mask = (1 << self.numSetBits) - 1
        mask = mask << self.numBlockOffsetBits
        return (self.addr & mask) >> self.numBlockOffsetBits


    def getOffset(self):
        """Returns the offset bits with bitmasking"""
        mask = (1 << self.numBlockOffsetBits) - 1
        return self.addr & mask


    def calcNumBlockOffsetBits(self):
        """Calculate and return the number of offset bits"""
        return int(math.ceil(math.log(float(self.numBlocks), 2)))


    def calcNumSetBits(self):
        """Calculate and return the number of set bits"""
        return int(math.ceil(math.log(float(self.numSets), 2)))


    def calcNumTagBits(self):
        """Calculate and return the number of tag bits"""
        return self.addrBits - self.numBlockOffsetBits - self.numSetBits

File: p1/src_c++/address.cpp
/* Michael LeMay, 1/13/19 */


#include <iostream>
#include <cmath>
#include "address.h"


/* Address implementation */

/*
 * Basic constructor
 */
Address::Address(int newAddr, int newAddrBits, int newNumSets, int newNumBlocks) {
  addr = newAddr;
  addrBits = newAddrBits;
  numSets = newNumSets;
  numBlocks = newNumBlocks;

  /*Calculate the number of bits*/
  calcNumBlockOffsetBits();
  calcNumSetBits();
  calcNumTagBits();
}


/*
 * Grabs first numBits bits of the address.
 * numBits must be a valid number of bits to request
 */
int Address::getTag() {
  /*Creates the mask of correct size*/
//  std::cout << "ex: "  << ~0 << ", " << ~0 << (sizeof(int) - numBits);
  int mask = (1 <<  numTagBits) - 1;
  mask = mask << (numSetBits + numBlockOffsetBits);
  printf("mask: %d", mask);
  return (addr & mask) >> (numSetBits + numBlockOffsetBits);
}


/*
 * Grabs middle set index from addr
 * numBits must be a valid number of bits to request
 */
int Address::getIndex() {
  /*Creates the mask of correct size*/
  int mask = (1 << numSetBits) - 1;
  /*Shifts mask over to cover the set index*/
  mask = mask << numBlockOffsetBits;
  return (addr & mask) >> numBlockOffsetBits;
}


/*
 * Grabs last numBits bits of addr.
 * numBits must be a valid number of bits to request
 */
int Address::getOffset() {
  /*Creates the mask of correct size*/
  int mask = (1 << numBlockOffsetBits) - 1;
  return addr & mask;
}


/*Calculate number of bits for each part of address, load them in object*/

/*
 * Calculates number of offset bits with log base 2 rounded up
 */
void Address::calcNumBlockOffsetBits() {
   numBlockOffsetBits = (int) ceil(log2((double) numBlocks));
}


/*
 * Calculate the number of set index bits with log base 2 rounded up
 */
void Address::calcNumSetBits() {
   numSetBits = (int) ceil(log2((double) numSets));
}


/*
 * Calculate the number of tag bits as the difference between addr bits
 * and the set bits and block offset bits
 */
void Address::calcNumTagBits() {
  numTagBits = addrBits - numBlockOffsetBits - numSetBits;
}
